src/services/completions.ts

```diff
@@ -2121,10 +2126,29 @@ export function getCompletionEntriesFromSymbols(
             }
             // Filter out variables from their own initializers
             // `const a = /* no 'a' here */`
-            if (variableDeclaration && symbol.valueDeclaration === variableDeclaration) {
+            if (tryCast(variableOrParameterDeclaration, isVariableDeclaration) && symbol.valueDeclaration === variableOrParameterDeclaration) {
                 return false;
             }
 
+            // Filter out parameters from their own initializers
+            // `function f(a = /* no 'a' and 'b' here */, b) { }` or
+            // `function f<T = /* no 'T' here */>(a: T) { }`
+            const symbolDeclaration = symbol.valueDeclaration ?? symbol.declarations?.[0];
+            if (variableOrParameterDeclaration && symbolDeclaration && (
+                (isTypeParameterDeclaration(variableOrParameterDeclaration) && isTypeParameterDeclaration(symbolDeclaration)) ||
+                (isParameter(variableOrParameterDeclaration) && isParameter(symbolDeclaration))
+            )) {
+                const symbolDeclarationPos = symbolDeclaration.pos;
+                const parameters = isParameter(variableOrParameterDeclaration) ?
+                    variableOrParameterDeclaration.parent.parameters :
+                    isInferTypeNode(variableOrParameterDeclaration.parent) ?
+                        undefined :
+                        variableOrParameterDeclaration.parent.typeParameters;
+                if (symbolDeclarationPos >= variableOrParameterDeclaration.pos && parameters && symbolDeclarationPos < parameters.end) {

```

Comment:
I don't understand any of these predicates.
- why not check `symbolDeclaration === variableOrParameterDeclaration`?
- why check `parameters`?
- when can symbolDeclaration.pos be larger than parameters.end?

Reply:
reading the previous reviews, it seems like this code was a suggestion from @DanielRosenwasser

Reply:
@DanielRosenwasser can you explain this check?

Reply:
Yeah - it seemed unnecessary to run through all the parameters to find the parameter index and compare with that. Just checking the textual position seemed like enough. So we just check if the symbol was declared *somewhere after* the current parameter, but still within the parameter list.

Reply:
Also, I'm pretty sure you can't really do a naive walk of the parameter list because of destructuring.

Reply:
> when can symbolDeclaration.pos be larger than parameters.end

Technically we could remove that check - you probably don't want to provide completions for *any* declaration even if it's within the parameter list.

Reply:
Well...wait, that's not true.

```ts
function f(x = /**/) {
}

var y;
```

You still want completions for `y` there. So that's what I was thinking.

Reply:
> You still want completions for `y` there. So that's what I was trying to model.

But we have the following check: `isParameter(variableOrParameterDeclaration) && isParameter(symbolDeclaration)` line: https://github.com/microsoft/TypeScript/pull/52690/files#diff-4237fd08a5f920543e184b7748bc8ae000d9b1c031810ba375b0053362ec2b6aR2139, since `y` is not parameter it doesn't get included anyway

Reply:
If what's there now works, I think it's fine. It's pretty understandable and more general than strict equality of declarations.